File: apps/csharp/src/Ch13/WordCount.cs
namespace GrokkingConcurrency.CSharp.Ch13;

/// <summary>
/// MapReduce パターンによるワードカウント
/// </summary>
public static class WordCount
{
    /// <summary>
    /// Map: テキストを (word, 1) のペアに変換
    /// </summary>
    public static List<KeyValuePair<string, int>> Map(string text)
    {
        return text.ToLower()
            .Split(' ', StringSplitOptions.RemoveEmptyEntries)
            .Select(word => new KeyValuePair<string, int>(word, 1))
            .ToList();
    }

    /// <summary>
    /// Reduce: 同じ単語のカウントを集約
    /// </summary>
    public static Dictionary<string, int> Reduce(List<KeyValuePair<string, int>> pairs)
    {
        return pairs
            .GroupBy(p => p.Key)
            .ToDictionary(g => g.Key, g => g.Sum(p => p.Value));
    }

    /// <summary>
    /// MapReduce でワードカウント
    /// </summary>
    public static Dictionary<string, int> CountWords(List<string> texts)
    {
        // Map フェーズ（並列）
        var mapped = texts
            .AsParallel()
            .SelectMany(text => Map(text))
            .ToList();

        // Reduce フェーズ
        return Reduce(mapped);
    }
}

File: apps/fsharp/src/Ch13/WordCount.fs
namespace GrokkingConcurrency.Ch13

open System

module WordCount =
    /// Map: Convert text to (word, 1) pairs
    let map (text: string) : (string * int) list =
        text.ToLower().Split([|' '; '\t'; '\n'; '\r'|], StringSplitOptions.RemoveEmptyEntries)
        |> Array.filter (fun s -> s.Length > 0)
        |> Array.map (fun word -> (word, 1))
        |> Array.toList

    /// Reduce: Aggregate word counts
    let reduce (pairs: (string * int) list) : Map<string, int> =
        pairs
        |> List.groupBy fst
        |> List.map (fun (key, group) -> (key, group |> List.sumBy snd))
        |> Map.ofList

    /// MapReduce: Count words in multiple texts using parallel execution
    let countWords (texts: string list) : Map<string, int> =
        // Map phase (parallel)
        let mapped =
            texts
            |> List.toArray
            |> Array.Parallel.collect (fun text -> map text |> List.toArray)
            |> Array.toList

        // Reduce phase
        reduce mapped
